File: __legacy/src/endpoints/auth/provider/github.py
import os, httpx
from urllib.parse import urlencode
from src import libs

class ProviderGithub:

    # set values
    name = 'github'
    type = 'OAuth'
    scope = 'user'
    header_type = 'Bearer'
    client_id = os.getenv('AUTH_GITHUB_CLIENT_ID')
    client_secret = os.getenv('AUTH_GITHUB_CLIENT_SECRET')

    # url path
    url_authorization = 'https://github.com/login/oauth/authorize'
    url_token = 'https://github.com/login/oauth/access_token'
    url_userinfo = 'https://api.github.com/user'

    def create_authorize_url(self, state: str):
        qs = urlencode({
            'client_id': self.client_id,
            'redirect_uri': f'{libs.url_path}/auth/callback/{self.name}/',
            'scope': self.scope,
            'state': state,
        })
        return f'{self.url_authorization}/?{qs}'

    async def get_token(self, code: str) -> dict | None:
        async with httpx.AsyncClient() as client:
            res = await client.post(
                url = self.url_token,
                headers = { 'Accept': 'application/json' },
                data = {
                    'client_id': self.client_id,
                    'client_secret': self.client_secret,
                    'code': code,
                    'redirect_uri': f'{libs.url_path}/auth/callback/{self.name}/',
                },
            )
            if res.status_code != 200: raise Exception(res.text)
            json = res.json()
        return {
            'access': json['access_token'],
            'expires': json['expires_in'],
            'refresh': json['refresh_token'],
        }

    async def get_user(self, token: str) -> dict | None:
        async with httpx.AsyncClient() as client:
            res = await client.get(
                url = self.url_userinfo,
                headers = { 'Authorization': f'{self.header_type} {token}' },
            )
            if res.status_code != 200: raise Exception(res.text)
            json = res.json()
        return {
            'id': json['id'],
            'name': json['name'],
            'email': json['email'],
            'avatar': json['avatar_url'],
        }

    def check_user_id(self, user_id: str, user_data: dict) -> bool:
        return str(user_id) == str(user_data['id'])

    async def renew_access_token(self, refresh_token: str) -> dict | None:
        if not refresh_token: return None
        async with httpx.AsyncClient() as client:
            res = await client.post(
                url = self.url_token,
                headers = { 'Accept': 'application/json' },
                data = {
                    'client_id': self.client_id,
                    'client_secret': self.client_secret,
                    'grant_type': 'refresh_token',
                    'refresh_token': refresh_token,
                },
            )
            if res.status_code != 200: raise Exception(res.text, res.status_code or 400)
            json = res.json()
        return {
            'access': json['access_token'],
            'expires': json['expires_in'],
            'refresh': json['refresh_token'],
        }

File: src/classes/provider/Github.ts
/**
 * ProviderGithub
 * 깃허브 인증 서비스 인터페이스 클래스
 */

import ServiceError from '@/classes/ServiceError'
import Provider from './Provider'
import { PROVIDER_CODE, PROVIDER_TYPE } from './'
import { parseQueryString } from '@/libs/strings'
import { PATHS } from '@/libs/assets'

export default class ProviderGithub extends Provider {

  public code = PROVIDER_CODE.GITHUB
  public type = PROVIDER_TYPE.OAUTH
  public description = `oAuth by ${PROVIDER_CODE.GITHUB}`
  private scope = 'user'
  private headerType = 'Bearer'
  private clientId = Bun.env.AUTH_GITHUB_CLIENT_ID
  private clientSecret = Bun.env.AUTH_GITHUB_CLIENT_SECRET
  private url_authorization = 'https://github.com/login/oauth/authorize'
  private url_token = 'https://github.com/login/oauth/access_token'
  private url_userinfo = 'https://api.github.com/user'

  constructor() {
    super()
  }

  static checkUserId(userId: string, userData: ZZ): boolean
  {
    return userId === userData.id
  }

  public createAuthorizeUrl(state: string): string
  {
    const _query = parseQueryString({
      client_id: this.clientId,
      redirect_uri: `${PATHS.URL}/auth/callback/${this.code}/`,
      scope: this.scope,
      state,
    })
    return `${this.url_authorization}?${_query}`
  }

  public async getToken(code: string): Promise<ZZ>
  {
    const res = await fetch(this.url_token, {
      method: 'post',
      headers: { 'Accept': 'application/json' },
      body: new URLSearchParams({
        client_id: this.clientId,
        client_secret: this.clientSecret,
        code,
        redirect_uri: `${PATHS.URL}/auth/callback/${this.code}/`,
      } as ZZ),
    })
    const _json: any = await res.json()
    if (!res?.ok)
    {
      throw new ServiceError('인증 서비스 토큰을 가져올 수 없습니다.', {
        status: 401,
        text: _json.error_description,
      })
    }
    if (!_json?.access_token)
    {
      throw new Error('응답 데이터에서 엑세스 토큰이 없습니다.')
    }
    return {
      access: _json.access_token,
      expires: _json.expires_in,
      refresh: _json.refresh_token,
    }
  }

  public async getUser(accessToken: string): Promise<ZZ>
  {
    const res = await fetch(this.url_userinfo, {
      method: 'get',
      headers: { 'Authorization': `${this.headerType} ${accessToken}` },
    })
    const _json = (await res.json()) as ZZ
    if (!_json?.id) throw new Error('응답 데이터가 없습니다.')
    return {
      id: _json.id,
      name: _json.name,
      email: _json.email,
      avatar: _json.avatar_url,
    }
  }

  public async renewToken(op: ZZ): Promise<ZZ|undefined>
  {
    if (!op.refresh) return
    const res = await fetch(this.url_token, {
      method: 'post',
      headers: { 'Accept': 'application/json' },
      body: new URLSearchParams({
        client_id: this.clientId,
        client_secret: this.clientSecret,
        grant_type: 'refresh_token',
        refresh_token: op.refresh,
      } as ZZ)
    })
    const _json: any = await res.json()
    if (!res?.ok)
    {
      throw new ServiceError('엑세스 토큰 재발급 실패', {
        status: 401,
        text: _json.error_description,
      })
    }
    return {
      access: _json.access_token,
      accessPublic: Provider.getPublicToken(_json.access_token),
      refresh: _json.refresh_token,
      expires: _json.expires_in,
    }
  }

}
